sdks/python/apache_beam/transforms/deduplicate.py

```diff
@@ -0,0 +1,133 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# pytype: skip-file
+
+"""a collection of ptransforms for deduplicating elements."""
+
+from __future__ import absolute_import
+from __future__ import division
+
+import typing
+
+from apache_beam import typehints
+from apache_beam.coders.coders import BooleanCoder
+from apache_beam.transforms import core
+from apache_beam.transforms import ptransform
+from apache_beam.transforms import userstate
+from apache_beam.transforms.timeutil import TimeDomain
+from apache_beam.utils import timestamp
+
+__all__ = [
+    'Deduplicate',
+    'DeduplicatePerKey',
+]
+
+K = typing.TypeVar('K')
+V = typing.TypeVar('V')
+
+
+@typehints.with_input_types(typing.Tuple[K, V])
+@typehints.with_output_types(typing.Tuple[K, V])
+class DeduplicatePerKey(ptransform.PTransform):
+  """ A PTransform which deduplicates <key, value> pair over a time domain and
+  threshold. Values in different windows will NOT be considered duplicates of
+  each other. Deduplication is best effort.
+
+  The durations specified may impose memory and/or storage requirements within
+  a runner and care might need to be used to ensure that the deduplication time
+  limit is long enough to remove duplicates but short enough to not cause
+  performance problems within a runner. Each runner may provide an optimized
+  implementation of their choice using the deduplication time domain and
+  threshold specified.
+
+  Does not preserve any order the input PCollection might have had.
+  """
+  def __init__(self, processing_time_duration=None, event_time_duration=None):
+    if processing_time_duration is None and event_time_duration is None:
+      raise ValueError(
+          'DeduplicatePerKey requires at lease provide either'
+          'processing_time_duration or event_time_duration.')
+    self.processing_time_duration = processing_time_duration
+    self.event_time_duration = event_time_duration
+
+  def _create_deduplicate_fn(self):
+    processing_timer_spec = userstate.TimerSpec(
+        'processing_timer', TimeDomain.REAL_TIME)
+    event_timer_spec = userstate.TimerSpec('event_timer', TimeDomain.WATERMARK)
+    state_spec = userstate.BagStateSpec('seen', BooleanCoder())
```

Comment:
Perhaps this should be a combining state?

Reply:
The `seen_state` is only set once per key during that duration. I'm not sure whether combining state is more suitable here. What do you think?

Reply:
To follow up from what we discussed off-line, I think combining state semantically makes slightly more sense here, but as we're not adding to the state unless it's the first element there's no performance concerns, and I'm also fine with this as is.